tokio/src/time/sleep.rs

```diff
@@ -353,6 +353,26 @@ impl Sleep {
         self.reset_inner(deadline)
     }
 
+    /// Resets the `Sleep` instance to a new deadline without reregistering it
+    /// to be woken up.
+    ///
+    /// Calling this function allows changing the instant at which the `Sleep`
+    /// future completes without having to create new associated state and
+    /// without having it registered. This is required in e.g. the
+    /// [crate::time::Interval] where we want to reset the internal [Sleep]
+    /// without having it wake up the last task that polled it.
+    ///
+    /// This function can be called both before and after the future has
+    /// completed.
+    ///
+    /// To call this method, you will usually combine the call with
+    /// [`Pin::as_mut`], which lets you call the method without consuming the
+    /// `Sleep` itself.
+    pub fn reset_without_reregister(self: Pin<&mut Self>, deadline: Instant) {
```

Comment:
Adding a new public function should be a high bar. Either the existing behavior is incorrect, and we should fix it or accept it the way it is and evaluate adding a new API.

My concern with `reset_without_reregister` is that it makes the caller decide. Which variant should they use? Choosing one requires knowing about "waking," which isn't required for using Tokio as a beginner.

Reply:
Hm that is a good point  :thinking: I do think that users of `tokio::time::Sleep` would always want the re-registering variant. The non-re-registering is only needed for the specific use case in `tokio::time::Interval`. I changed the visibility to `pub(crate)`. Would that be OK?